lib: os: p4wq: add a done handler and use it to free up memory

FlorianWeber1018 opened the pull request:
This commit fixes the bug that the memory of the work request
is freed in the work handler and before it is removed from the p4wq.
This has leaded to a fatal error if the rtio i2c fallback is used frequently enough.
To fix this an additional handler is added to give the submitting code
(e.g. rtio workq) a possibility to free up dynamically allocated memory,
when it is not needed anymore by the p4wq.
This is now used in the rtio workq.

Fixes #84384

andyross reviewed (changes requested):
This looks like a deadlock to me.  The handler is being called with the spinlock held, which means that if it tries to do anything else with the queue (a... pretty obvious use case if you think about it!) it's going to try to take the lock again and hang.

It's **probably** sufficient to just release the lock around the call, just as is done for the handler.

ubieda commented on subsys/rtio/rtio_workq.c:
```diff
@@ -28,7 +28,13 @@ static void rtio_work_handler(struct k_p4wq_work *work)
 	req->handler(iodev_sqe);
+}
 
+static void rtio_done_handler(struct k_p4wq_work *work)
```
```suggestion
static void rtio_work_req_done_handler(struct k_p4wq_work *work)
```

ubieda reviewed (changes requested):
Thanks for this fix. I suggest splitting in separate patches:
- P4WQ changes.
- RTIO Workqueue changes.

Also, let's fix compliance issues.

FlorianWeber1018 commented:
> Thanks for this fix. I suggest splitting in separate patches:
> 
> * P4WQ changes.
> * RTIO Workqueue changes.
> 
> Also, let's fix compliance issues.

i have not splitted this i separate commits because only changing the p4wq leads to the callback with random content because it has to be initialized there.

andyross reviewed (commented):
And a design note: this problem is almost exactly isomorphic to the one solved in commit 56610bdafb4d8f506218efa8270799301f795d74, which added the "done_sem" field which is given after the last access to the item.

Is there a reason you can't use that?  Seems to me like we want one "done" notification mechanism and not two.

Also, if you absolutely need the callback[1], maybe consider putting it onto the main queue object and not duplicating it for every item (and having it take the item pointer as an argument, obviously).  It strikes me as very rare that you'd want these to actually be different functions for different items in the same queue, and even if they were you could handle that with an if() inside the single handler.

[1] In which case we should look at deprecating `done_sem` and migrating the existing users to the new scheme.

FlorianWeber1018 commented:
> And a design note: this problem is almost exactly isomorphic to the one solved in commit [56610bd](https://github.com/zephyrproject-rtos/zephyr/commit/56610bdafb4d8f506218efa8270799301f795d74), which added the "done_sem" field which is given after the last access to the item.
> 
> Is there a reason you can't use that? Seems to me like we want one "done" notification mechanism and not two.
> 
> Also, if you absolutely need the callback[1], maybe consider putting it onto the main queue object and not duplicating it for every item (and having it take the item pointer as an argument, obviously). It strikes me as very rare that you'd want these to actually be different functions for different items in the same queue, and even if they were you could handle that with an if() inside the single handler.
> 
> [1] In which case we should look at deprecating `done_sem` and migrating the existing users to the new scheme.

Please correct me if I am wrong but if i use the done_sem for waiting for completion in the handler, is this not a deadlock?
And if I am waiting in the submit function it is not asynchronous anymore. So from my understanding it is the only implementaion I have in mind, but I am open for proposals.

andyross commented:
No, you'd use some other thread context to reap your workers, like e.g. polling on all the done_sem's, etc...  How that fits into your architecture is an app thing, and not all tools work perfectly.

I'm just saying that @lyakh saw and addressed seemingly the same issue already, and we should have a software system with one solution to a problem and not two.

andyross commented:
FWIW, I grepped around, and can't find any users in-tree (well, rtio initializes the thing but doesn't use it) in either Zephyr or SOF.  So IMHO simply deprecating/removing the older mechanism would be acceptable unless @teburd or @lyakh know of other uses.  But do consider making it a queue-global function pointer and not per-item.  Work items should be small and lean.

ubieda commented:
> But do consider making it a queue-global function pointer and not per-item. Work items should be small and lean.

Just to add more context on this: The use-case where RTIO workqueue needs this is basically because there's a Mem-Slab with RTIO work items, where the P4WQ item itself is contained:

https://github.com/zephyrproject-rtos/zephyr/blob/1d441c239e325b16fa3b8e4810119eca69e337cb/subsys/rtio/rtio_workq.c#L18-L21

https://github.com/zephyrproject-rtos/zephyr/blob/1d441c239e325b16fa3b8e4810119eca69e337cb/include/zephyr/rtio/work.h#L32-L34

The issue we're facing is that the RTIO Workqueue needs to put the item back in the pool, but there's no "reaper" context waiting to synchronize with this operation (at least not one that is exposed to the RTIO Workqueue APIs, it just knows that is RTIO). Right now the RTIO handler is putting the item back in the pool before exiting, but that is inherently racey as the P4WQ needs to finalize the item before someone else takes it and tries to submit it.

https://github.com/zephyrproject-rtos/zephyr/blob/1d441c239e325b16fa3b8e4810119eca69e337cb/subsys/rtio/rtio_workq.c#L23-L33

FlorianWeber1018 commented:
@ubieda thanks for adding the context for me as a non native speaker. You have perfectly described the issue.
@andyross from my knowledge there is no argument against storing the function pointer globally per p4wq so i am fine with doing that if there is a consense to do that. I was also surprised that the rtio_workq and the test for the p4wq are the only thing in tree which uses the p4wq at all.

FlorianWeber1018 committed:
rtio: workq: bugfix of memory allocation

This commit fixes the bug that the memory of the work request is
freed up in the work handler, before it is not needed anymore by the p4wq.
This is fixed now, by using the new done_handler in the p4wq
for freeing up that memory.

Signed-off-by: Florian Weber <Florian.Weber@live.de>

FlorianWeber1018 commented:
@andyross I have moved the handler now to the k_p4wq. Additionally i have added new macros for static initialisation with the new done_handler and when the old macro is used the pointer is initialized with null. So the code is be backwards compatible to existing code at this point. 
I have not found a way to do this in an elegant way in case of the non static initialization because it is unclear to me how to use the p4wq that way. Just calling k_p4wq_init seems to be not sufficent to me, as it does not set the flags as it is implemented in case of the static initialization. Could you please help me understanding this? If it´s in the response of the application to set the flags in that case to a valid value that could also be the rule for the new function pointer right?

@ubieda because of the backwards compatibility in case of static initialization I was able to separate the changes in two commits as required.

What is left now, is the question if we want to deprecate or remove the old mechanism for detecting when its done with the semophore. I agree that it is not good to have two redundant mechanisms.

andyross commented:
> I have not found a way to do this in an elegant way in case of the non static initialization 

Just a `p4wq_set_done_handler()` should be fine IMHO.  If there are bugs in init, FWIW, that's a test coverage failure.

andyross reviewed (dismissed):
Oh, but remove -1.  This looks clean to me.  We should definitely think about deprecating the semaphore (which apps could implement on their own with this handler).

ubieda reviewed (dismissed):
Stamping with +1 assuming CI red checks will be addressed. Thanks for the fix!

lyakh commented:
> FWIW, I grepped around, and can't find any users in-tree (well, rtio initializes the thing but doesn't use it) in either Zephyr or SOF. So IMHO simply deprecating/removing the older mechanism would be acceptable unless @teburd or @lyakh know of other uses. But do consider making it a queue-global function pointer and not per-item. Work items should be small and lean.

@andyross sorry, you couldn't find users of what exactly? If you meant that my fix, then I think it's still used by SOF https://github.com/thesofproject/sof/blob/e95d7ee95827cd2fae2565c0c92b039fc97a6605/src/idc/zephyr_idc.c#L168

FlorianWeber1018 commented:
@lyakh yes we meant your fix. The problem is that it does not fit to our problem we have in the RTIO workq where we can´t just wait for completion because it does not fit to the asynchronous nature of the problem. 

At the moment we consider to deprecate your fix because with my implementation the semaphore could also implemented in the user code, if needed and we don´t want to have two redundant mechanism for two really similar problems (signaling when its done).

@andyross I see one problem: at the moment the handler is not called when the work has been canceled because I thought if the user has canceled the work he knows because of the return value how to proceed, but a implementation of the done_semaphore in the user land is not that straight forward in that case. 
What do you think should the done_handler also be called if the work was canceled before execution?

lyakh commented:
> At the moment we consider to deprecate your fix because with my implementation the semaphore could also implemented in the user code, if needed and we don´t want to have two redundant mechanism for two really similar problems (signaling when its done).

@FlorianWeber1018 @andyross right, we can deprecate the completion semaphore and move SOF to also use the callback, but yes, we'd need some time to migrate, so we shouldn't just remove it, but wait until SOF has updated and moved over to a new enough Zephyr version

FlorianWeber1018 commented on lib/os/p4wq.c:
```diff
@@ -105,6 +105,14 @@ static FUNC_NORETURN void p4wq_loop(void *p0, void *p1, void *p2)
+				k_spin_unlock(&queue->lock, k);
+
+				if (queue->done_handler) {
+					queue->done_handler(w);
```
The semaphore is only to regain ownership of the work item and not of the whole queue right? Because the callback lives in the queue it is still in the ownership of the queue. If the semaphore is used (and in that case the done_handler must not be used) the done handler is still null and does not get called.

lyakh replied:
> The semaphore is only to regain ownership of the work item and not of the whole queue right?

Correct

> Because the callback lives in the queue it is still in the ownership of the queue. If the semaphore is used (and in that case the done_handler must not be used) the done handler is still null and does not get called.

Well, so far the callback isn't in the mainline yet and so it has no users. So, yes, it's perfectly safe to assume, that nobody uses both so far. But I'd prefer to also make that clear in the code, that nobody should do that - for as long as we have both in place. How about
```
				if (queue->done_handler) {
					k_spin_unlock(&queue->lock, k);
					queue->done_handler(w);
					k_spin_unlock(&queue->lock, k);
				} else {
					k_sem_give(&w->done_sem);
				}
```
and - btw - yes, no need to drop the spinlock if the callback isn't used. Let me give this a -1 - I think this deserves at least a short discussion, and sorry if you do end up making a new revision and temporarily losing all the approvals.

FlorianWeber1018 commented:
I think the best way is to mark the k_p4wq_wait function as deprecated and remove the stuff entirely after a good transition period.

ubieda commented:
I believe this bug-fix applies to both `4.0-branch` as well as `3.7-branch`. I suggest we capture an issue and add the respective labels.

andyross reviewed (dismissed):
I'm cool with this version, though we should still think on deprecation

lyakh reviewed (changes requested):
let's not drop and re-acquire the spinlock needlessly, see https://github.com/zephyrproject-rtos/zephyr/pull/84322#discussion_r1926488970

FlorianWeber1018 commented:
> let's not drop and re-acquire the spinlock needlessly, see [#84322 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/84322#discussion_r1926488970)

agreed.

FlorianWeber1018 committed:
lib: os: p4wq: add done handler

Add an optional handler to the p4wq to give the submitting code
(e.g. rtio workq) a possibility execute code after the work was
succesfully executed.

Signed-off-by: Florian Weber <Florian.Weber@live.de>

lyakh reviewed (approved):
Thanks, this version should work both ways - with a semaphore or a `.done()` callback

FlorianWeber1018 commented:
@andyross  could you please finally review this that it can be merged?
